core/Updates/2.14.0.php

```diff
@@ -0,0 +1,46 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\Updates;
+
+use Piwik\Config;
+use Piwik\Updater;
+use Piwik\Updates;
+use Piwik\Plugin\Manager;
+
+class Updates_2_14_0 extends Updates
+{
+    public function doUpdate(Updater $updater)
+    {
+        $this->uninstallPlugin('UserSettings');
+    }
+
+    private function uninstallPlugin($plugin)
+    {
+        $pluginManager = Manager::getInstance();
+
+        if ($pluginManager->isPluginInstalled($plugin)) {
+
+            if ($pluginManager->isPluginActivated($plugin)) {
+                $pluginManager->deactivatePlugin($plugin);
+            }
+
+            $pluginManager->unloadPlugin($plugin);
+            $pluginManager->uninstallPlugin($plugin);
+
+        } else {
+            $this->makeSurePluginIsRemovedFromFilesystem($plugin);
+        }
+    }
```

Comment:
Unfortunately, the design of `Plugin` and `Plugin\Manager` is a bit broken. I would probably have expected to have the code of `Plugin\Manager::uninstall` in `Plugin::uninstall` and The code above in `Plugin\Manager::uninstall`. There are some reasons though why it can and should be as it currently is. I was thinking about adding this code to another class such as the Updater so one can do `$this->uninstallPlugin()` in the future but I preferred to see if it will be used somewhere else again (we had such a use case already here: https://github.com/piwik/piwik/blob/2.13.1/core/Updates/2.4.0-b3.php ) and rather then refactor it. We could change the behaviour of `Plugin\Manager::uninstall` though I reckon.